### PR TITLE
fix: bugs on border cases of udiv_mod

### PR DESCRIPTION
**Pull request by Pablo-Dallegri**
# Description

## Problem\*

Failures when attempting to prove certain `udiv_mod` operations.

## Summary\*
:beetle: Unsigned integer division fails if the numerator can be represented with less bits than the divisor, due to an underflow on the subtraction at:
https://github.com/noir-lang/noir-bignum/blob/1fcb2c91a9a471e511f313c15f463c3c6f987f14/src/fns/unconstrained_ops.nr#L174

:lady_beetle: Unsigned integer division also fails in cases were the least significant limb of the remainder is 2¹²⁰-1, due to a miscalculation of borrow flags, halting the proving at:
https://github.com/noir-lang/noir-bignum/blob/1fcb2c91a9a471e511f313c15f463c3c6f987f14/src/fns/constrained_ops.nr#L463

:wrench: This PR adds tests for both cases, and proposes fixes.

## Additional Context

These bugs where found while developing in the context of [NRG#2](https://github.com/orgs/noir-lang/discussions/6359), and tests were reduced to minimal examples.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: src/tests/bignum_test.nr [hunk 1]
```diff
@@ -632,6 +632,21 @@ fn test_udiv_mod_U256() {
...
+
+#[test]
+fn test_20_udiv_mod_11() {
+    let _1: U256 = BigNum::one();
+    let _2_POW_120: U256 = BigNum::from_slice([0, 1, 0]);
+    let _2_POW_121: U256 = BigNum::from_slice([0, 2, 0]);
+    assert(_2_POW_121.udiv_mod(_2_POW_120 + _1) == (_1, _2_POW_120 - _1));
+}
+
 // // N.B. witness generation times make these tests take ~15 minutes each! Uncomment at your peril
```
Comment (kashbrti) [inline src/tests/bignum_test.nr:650]: Thanks a lot for this! could you please add a test that would fail in the edge case? You can use the `should_fail_with` flag for this.

**Review by kashbrti — CHANGES_REQUESTED** [review]

File: src/fns/unconstrained_ops.nr [hunk 2]
```diff
@@ -171,34 +171,38 @@ pub(crate) unconstrained fn __udiv_mod<let N: u32>(
     let mut divisor_u60: U60Repr<N, 2> = U60Repr::from(divisor);
     let b = divisor_u60;
 
-    let mut bit_difference = remainder_u60.get_msb() - divisor_u60.get_msb();
```
Comment (kashbrti) [inline src/fns/unconstrained_ops.nr:174]: I see how this line can cause issues when the denominator has more bits in its u60 representation. Thanks for catching this!

File: src/fns/unconstrained_ops.nr [hunk 3]
```diff
@@ -171,34 +171,38 @@ pub(crate) unconstrained fn __udiv_mod<let N: u32>(
...
+    if !remainder_u60.gte(divisor_u60) {
```
Comment (kashbrti) [inline src/fns/unconstrained_ops.nr:174]: I wonder if there would be a better way to catch this, without using the if statement. @TomAFrench would it be more efficient to look at the index where the `msb` of the denominator is rather than using the if statement? 
regardless the change should fix this issue.

File: src/fns/unconstrained_ops.nr [hunk 4]
```diff
@@ -171,34 +171,38 @@ pub(crate) unconstrained fn __udiv_mod<let N: u32>(
...
+        accumulator_u60 = accumulator_u60.shl(bit_difference);
 
-    if (divisor_u60.gte(remainder_u60 + U60Repr::one())) {
-        divisor_u60.shr1();
-        accumulator_u60.shr1();
-    }
-    for _ in 0..(N * 120) {
-        if (remainder_u60.gte(b) == false) {
-            break;
+        if (divisor_u60.gte(remainder_u60 + U60Repr::one())) {
```
[the remark this answers is not in the document]
Comment (TomAFrench) [reply src/fns/unconstrained_ops.nr:183]: They're from the if statement causing extra indentation.

File: src/fns/unconstrained_ops.nr (hunk 3 above)
Comment (kashbrti) [reply src/fns/unconstrained_ops.nr:174]: @Pablo-Dallegri would you like to tackle this yourself or should I try to make the change before merging the PR?

File: src/fns/unconstrained_helpers.nr [hunk 5]
```diff
@@ -39,12 +39,11 @@ pub(crate) unconstrained fn __validate_gt_remainder<let N: u32>(
     let mut b_u60: U60Repr<N, 2> = U60Repr::from(rhs);
 
     let underflow = b_u60.gte(a_u60);
-    b_u60 += U60Repr::one();
     assert(underflow == false, "BigNum::validate_gt check fails");
     let mut result_u60: U60Repr<N, 2> = U60Repr { limbs: [0; 2 * N] };
 
     let mut carry_in: u64 = 0;
-    let mut borrow_in: u64 = 0;
+    let mut borrow_in: u64 = 1;
```
Comment (kashbrti) [inline src/fns/unconstrained_helpers.nr:46]: this removes the need of adding 1 to `b_u60`

File: src/fns/unconstrained_ops.nr [hunk 6]
```diff
@@ -171,34 +171,40 @@ pub(crate) unconstrained fn __udiv_mod<let N: u32>(
...
-    if (divisor_u60.gte(remainder_u60 + U60Repr::one())) {
-        divisor_u60.shr1();
-        accumulator_u60.shr1();
-    }
-    for _ in 0..(N * 120) {
-        if (remainder_u60.gte(b) == false) {
-            break;
+    let numerator_msb = remainder_u60.get_msb();
+    let divisor_msb = divisor_u60.get_msb();
+    if divisor_msb > numerator_msb {
```
Comment (kashbrti) [inline src/fns/unconstrained_ops.nr:176]: checks if the divisor is larger than the numerator, just returns the numerator as the remainder and the quotient should be zero. so this looks good.

File: src/tests/bignum_test.nr [hunk 7]
```diff
@@ -632,6 +632,21 @@ fn test_udiv_mod_U256() {
...
 }
 
+#[test]
+fn test_1_udiv_mod_2() {
+    let _0: U256 = BigNum::new();
+    let _1: U256 = BigNum::one();
+    assert(_1.udiv_mod(_1 + _1) == (_0, _1));
+}
+
+#[test]
```
Comment (kashbrti) [inline src/tests/bignum_test.nr:642]: Tests for catching unwanted edge case behaviour have also been added. the fix looks good to me.

**Review by kashbrti — APPROVED** [review]

**Comment by kashbrti** [comment]
@TomAFrench the CI seems to not be running on this PR for some reason. Is it because it's getting merged from a fork?

File: src/fns/unconstrained_ops.nr [hunk 8]
```diff
@@ -171,34 +171,40 @@ pub(crate) unconstrained fn __udiv_mod<let N: u32>(
...
+        ([0; N], numerator)
+    } else {
+        let mut bit_difference = numerator_msb - divisor_msb;
+
+        let mut accumulator_u60: U60Repr<N, 2> = U60Repr::one();
+        divisor_u60 = divisor_u60.shl(bit_difference);
+        accumulator_u60 = accumulator_u60.shl(bit_difference);
+
+        if (divisor_u60.gte(remainder_u60 + U60Repr::one())) {
```
Comment (TomAFrench) [inline src/fns/unconstrained_ops.nr:185]:
```suggestion
        // divisor > remainder
        if (!remainder_u60.gte(divisor_u60)) {
```

lhs >= rhs +1 => lhs > rhs => !(lhs <= rhs) => !(rhs >= lhs)

This is more efficient.

**Comment by kashbrti** [comment]
this is a PR from fork so I think @Pablo-Dallegri has to sync his fork for that.

**Comment by TomAFrench** [comment]
As a maintainer of the upstream repo you can push onto the PR branch. I'd recommend using the github extension in vscode to check out the PR branch and push a merge commit.

**Comment by kashbrti** [comment]
moved this PR to #128 to fix merging issues and CI not running so will close this.
